### PR TITLE
Add $hidden and $wdesktop

### PR DESCRIPTION
**Pull request by liskin**
This PR adds recording of two additional window attributes:

 * `$hidden`: true if the window is minimized or on a virtual desktop
   that isn't currently being shown on any screen

   (not named `$visible` as the window can still be occluded by other
   windows)

 * `$wdesktop`: which desktop the window belongs to

This makes it possible to write rules such as: if the browser is
currently active, but windows on another screen indicate I'm working on
project X, let's tag this as project X too.

Recording the desktop of individual windows also makes it easier to
analyze uncategorized samples. An example of what that looks like:

```
$ stack exec -- arbtt-capture --dump
2021-03-01 20:21:22 (192ms inactive):
    Current Desktop: 5:arbtt:tmp
    (.) [1:irc]        urxvt:          t[N] 3,5,7
    ( ) [1:irc]        app.element.io: Element
    ( ) [1:irc]        discord.com__app: #readme
    (.) [1:irc]        x-terminal-emulator: t[m] m[N] =INBOX: 0N 15O 3I
    (.) [1:irc]        x-terminal-emulator: index.wiki (~/taskwiki) - VIM
    ( ) [2:web]        liferea:        Liferea
    ( ) [2:web]        google-chrome:  adhd - Google Chrome
    ( ) [2:web]        google-chrome:  chrome://newtab - Google Chrome
    ( ) [3:.xmonad]    urxvt:          tomi@notes: /home/tomi/.xmonad
    ( ) [3:.xmonad]    urxvt:          tomi@notes: /home/tomi/.xmonad
    ( ) [3:.xmonad]    urxvt:          t[2] xmonad.hs (~/.xmonad) - VIM
    ( ) [4:arbtt]      urxvt:          tomi@notes: /home/tomi
    ( ) [4:arbtt]      urxvt:          tomi@notes: /home/tomi
    ( ) [4:arbtt]      urxvt:          categorize.m4 (~/.arbtt) - VIM
    (.) [5:arbtt:tmp]  urxvt:          t[13] PULLREQ_EDITMSG (~/src-haskell/arbtt/.git/worktrees/arbtt-tmp) - VIM
    (.) [5:arbtt:tmp]  urxvt:          59 warnings - arbtt-tmp
    (*) [5:arbtt:tmp]  urxvt:          tomi@notes: /home/tomi/src-haskell/arbtt-tmp
    ( ) [6:arbtt:web]  google-chrome:  Test.Tasty.Golden
    ( ) [12:_watch]    urxvt:          mc [tomi@notes]:~/download/_watch/Handpicked Adventure
```

Windows that are neither active nor hidden are now marked with `(.)`.

---

#### [Relax bounds in ls_put for Text (allow 2 more strings)](../commit/41f32cad9059a72eea39b3bcd0dc9bf78df00fd6)

I believe i < 255 implies succ i < 256, which still fits in Word8.
Or is that meant for future extensions?

(Might be a good idea to dedup strs, but doing that in
backward-compatible way is quite an ugly change.)

#### [Derive some previously handwritten instances](../commit/45034759fe3cd43e7c7b1d12eeb1bdf86f51d647)

It's been a while since GHC can do this, let's make the code shorter and
easier to change.

#### [Fix some deprecation warnings](../commit/6db94d847357d00cb86764a25ee457c137b512e4)

compiledWithXScreenSaver is deprecated and always True since X11-1.9

Control.Monad.Instances is deprecated since GHC 7.8

OverlappingInstances doesn't seem to be needed in StringRef at all.

Related: bb179f327677 ("Require X11 1.9")

#### [Refactor 3-tuple into data WindowData](../commit/712b505729ae1932263959ab0bdd870b683c3dde)

Prepare the code for adding more fields like visibility and desktop.

#### [Extend WindowData with wHidden and wDesktop](../commit/982f09f9ac4ab18605a14e30256310f2771ab100)

Unfortunately now every window with desktop info consumes 3 strings in
ListOfStringable, so we can only handle 85 concurrent windows instead of
127 before StringRef breaks down. Deduplication should help here, is it
worth the code complexity for backward compat, though?

#### [Make the new "window" and "desktop" fields optional in JSON import](../commit/9954266738f40a61135821005fa8f0d443be5a5f)

Probably useless for most people as the binary format is backwards
compatible, but as I now have two conflicting extensions of it (this and
the screensaver branch), migrating via JSON is safest, so this helps.

#### [Store small numbers (list/string lengths) as one byte](../commit/a5702777fd62d0b464d00fa49f26dfb087bca3ed)

This compensates for the log file enlargement caused by adding
additional fields, and actually makes the log smaller than it was
before.

#### [Add dump/recover tests for CaptureData binary versions 1-5](../commit/bc5ce7d70ab677295789893188755b2653072af2)

tests/binversions.log was crafted using newtypes, manually written
instances and manually verifying that the intermediate outputs work in
older versions of arbtt.

tests/binversions_borked.log is almost the same, just with extra bytes
in between records.

#### [Add stats/categorize test for $hidden and $wdesktop](../commit/e2e9e16798103a39bc15618071af591651441b79)

---

This isn't implemented for Win32 and OSX, as we don't support virtual
desktops there anyway and nobody seems to mind, and it took long enough
for me to do this already.

Works well with xmonad and GNOME, and I hope @thorgrin (a friend who
told me he'd also like to have this extra functionality) will test it on
KDE soon. :-)

### CODE REVIEW

**Comment by nomeata** [comment]
Fun stuff! Also great path sequence explanation :-)

In the example that you gave I guess the `(.)` windows are those on the (in your case) two workspaces visible on your (I assme) two screens? Or how else can windows from two different virtual desktops be visible?

File: src/Capture/OSX.hs [hunk 1]
```diff
@@ -16,9 +16,12 @@ captureData = do
...
 
-        let winData = map (
-                \(h,t,p) -> (h == foreground, T.pack t, T.pack p)
-                ) titles
+        let winData = do
+                (h, t, p) <- titles
+                return def{ wActive = h == foreground
+                          , wTitle = T.pack t
+                          , wProgram = T.pack p
+                          }
```
Comment (nomeata) [inline src/Capture/OSX.hs:24]: This is lists? Then I usually find list comprehensions nicer to read (and makes it easier to see that these indeed are just lists and not some fancy monad)

File: src/Capture/Win32.hs [hunk 2]
```diff
@@ -20,9 +20,12 @@ captureData = do
...
 
-        let winData = map (
-                \(h,t,p) -> (h == foreground, T.pack t, T.pack p)
-                ) titles
+        let winData = do
+                (h, t, p) <- titles
+                return def{ wActive = h == foreground
+                          , wTitle = T.pack t
+                          , wProgram = T.pack p
+                          }
```
Comment (nomeata) [inline src/Capture/Win32.hs:28]: dito, list comprehensions preferred

File: src/Capture/X11.hs [hunk 3]
```diff
@@ -60,10 +61,10 @@ captureData = do
...
         fwin <- followTreeUntil dpy (`elem` wins) fsubwin
 
-        winData <- forM wins $ \w -> (,,)
-            (w == fwin) <$>
-            (T.pack <$> getWindowTitle dpy w) <*>
-            (T.pack <$> getProgramName dpy w)
+        winData <- forM wins $ \w -> do
+            title <- T.pack <$> getWindowTitle dpy w
+            program <- T.pack <$> getProgramName dpy w
+            return def{ wActive = w == fwin, wTitle = title, wProgram = program }
```
Comment (nomeata) [inline src/Capture/X11.hs]: Did you consider using `NamedFieldPuns` here?

Also slightly worried about the `def` here.  This is a program, not a library, so it’s ok (actuall good!) to have to adjust all  places where this value is is creatd. So maybe use  a plain record literal here?

File: src/Data.hs [hunk 4]
```diff
@@ -96,23 +106,30 @@ instance Binary UTCTime where
...
 
 instance ListOfStringable CaptureData where
-  listOfStrings = concatMap (\(b,t,p) -> [t,p]) . cWindows
+  listOfStrings = concatMap (\wd -> [wTitle wd, wProgram wd]) . cWindows
 
 instance StringReferencingBinary CaptureData where
 -- Versions:
 -- 1 First version
 -- 2 Using ListOfStringable
+-- 3 Add cDesktop
```
Comment (nomeata) [inline src/Data.hs:123]: Thanks :-)

File: src/Data.hs [hunk 5]
```diff
@@ -125,11 +133,33 @@ instance StringReferencingBinary CaptureData where
...
                       "You can try to recover your data using arbtt-recover."
 
 fromWindowsV0 :: [(Bool, Text, Text)] -> [WindowData]
 fromWindowsV0 = map $ \(a,t,p) -> def{ wActive = a, wTitle = t, wProgram = p }
 
-toWindowsV0 :: [WindowData] -> [(Bool, Text, Text)]
-toWindowsV0 = map $ \wd -> (wActive wd, wTitle wd, wProgram wd)
+instance StringReferencingBinary WindowData where
+-- Versions:
+-- 0 3-tuple without version tag
```
Comment (nomeata) [inline src/Data.hs:149]: Add
> …, handled in `instance StringReferencingBinary CaptureData`

File: src/Data.hs [hunk 6]
```diff
@@ -39,13 +40,15 @@ data CaptureData = CaptureData
...
+        , wHidden :: Bool
         , wTitle :: Text
         , wProgram :: Text
+        , wDesktop :: Text
         }
   deriving (Show, Read, Generic, NFData)
 
 instance Default WindowData where
- def = WindowData False "" ""
+ def = WindowData False False "" "" ""
```
Comment (nomeata) [inline src/Data.hs:52]: I wonder if `hidden = True` might not be a better default for old data? Not sure, though.

File: src/Data.hs [hunk 7]
```diff
@@ -121,9 +122,10 @@ instance StringReferencingBinary CaptureData where
 -- 2 Using ListOfStringable
 -- 3 Add cDesktop
 -- 4 WindowData instead of 3-tuple
+-- 5 CompactNum
```
Comment (nomeata) [inline src/Data.hs]: Worth doing both with the same tag? Or do you already have data at tag 4 that you need to preserve?

File: src/Data/Binary/StringRef.hs [hunk 8]
```diff
@@ -47,10 +49,28 @@ instance (StringReferencingBinary a, StringReferencingBinary b, StringReferencin
...
 
+newtype CompactNum a = CompactNum { fromCompactNum :: a }
+
+instance (Integral a, Num a, Binary a) => StringReferencingBinary (CompactNum a) where
+    ls_put _ (CompactNum i)
+          | 0 <= i && i < 255 = putWord8 (fromIntegral i)
+          | otherwise         = putWord8 255 >> put i
+    ls_get _ = fmap CompactNum $ getWord8 >>= \case
+        i | 0 <= i && i < 255 -> return (fromIntegral i)
+          | otherwise         -> get
```
Comment (nomeata) [inline src/Data/Binary/StringRef.hs:60]: Did you consider LEB1238-encoding here? Same until 128, worse for `[128…255]`, but then again better.

**Review by nomeata — COMMENTED** [review]
Amazing! This is really old code (>10 years) that you are touching, and I knew so littele about good coding back then, so it’s quite a feat of you to find your way around, and make such changes so well :-)

I have a few questions/comments/suggestions, but nothing major.

**Comment by liskin** [comment]
> In the example that you gave I guess the `(.)` windows are those on the (in your case) two workspaces visible on your (I assme) two screens? Or how else can windows from two different virtual desktops be visible?

Yeah, exactly. In xmonad (and I think a few other tiling wms as well), when there are multiple screens, multiple virtual desktops are being shown. In GNOME, switching to another desktop switches workspace on both (all) screens.

File: src/Capture/OSX.hs (hunk 1 above)
Comment (liskin) [reply src/Capture/OSX.hs:24]: Yeah, I was trying to indent this sensibly and totally forgot about comprehensions. Dunno what was going on in my head back then. I'll adjust this. :-)

File: src/Capture/X11.hs (hunk 3 above)
Comment (liskin) [reply src/Capture/X11.hs]: I think I tried it and ghc told me I can't use them with def{…}. I specifically wanted to _not_ have to adjust Capture.OSX/Win32, but you're right that here it probably does make sense to use that.

File: src/Data.hs (hunk 7 above)
Comment (liskin) [reply src/Data.hs]: I don't, just wanted to have separate commits with tests still passing in between and so on. I can try to squash this somehow.

File: src/Data/Binary/StringRef.hs (hunk 8 above)
Comment (liskin) [reply src/Data/Binary/StringRef.hs:60]: Never heard of it. Not sure if worth the extra complexity though. I expect that in practice most numbers in arbtt logs are quite small, but then I have no idea what other people's environments look like. Do you think it's worth it? Shall I look into it?

File: src/Data.hs (hunk 6 above)
Comment (liskin) [reply src/Data.hs:52]: For the rules that I'm going to write using this new feature, it very well might be, yeah. Perhaps `hidden = not active` would be even better. It's still wrong, but at least it won't confuse rules that take into account visible windows.

Comment (nomeata) [reply src/Data.hs:52]: 👍🏻 to `hidden = not active`

File: src/Data.hs (hunk 7 above)
Comment (nomeata) [reply src/Data.hs]: Yup, no need to deal with decoding a tag 4 that never existed.

File: src/Data/Binary/StringRef.hs (hunk 8 above)
Comment (nomeata) [reply src/Data/Binary/StringRef.hs:60]: https://hackage.haskell.org/package/leb128-cereal is a package that I wrote for that a while ago, but we are using a different parsing library here (not `cereal`). Yeah, let’s not bother.

**Comment by liskin** [comment]
I think I've addressed your comments. Let me know when you've seen the fixups, I'll rebase autosquash them in their respective commits.

**Comment by nomeata** [comment]
Thanks, looks great!

**Comment by liskin** [comment]
Rebased. :-)

**Comment by nomeata** [comment]
Do you benefit from an immediate release, or can it just go into whenever the next one happens?

**Comment by liskin** [comment]
No need to release for me, I run this from git anyway (with the screensaver branch).

I'm also thinking of making another PR to add something like arbtt-filter, a command that reads/writes the binary log and applies a filter, to make it easier to rotate the log (I do look at the data, sometimes more than once a day, so I'd like to keep a month-sized chunks at most to keep arbtt-stats under a second).

**Comment by nomeata** [comment]
> I'm also thinking of making another PR to add something like arbtt-filter, a command that reads/writes the binary log and applies a filter, to make it easier to rotate the log (I do look at the data, sometimes more than once a day, so I'd like to keep a month-sized chunks at most to keep arbtt-stats under a second).

Sounds reasonable, I would accept that. Or maybe just add `--output-format=binary` to `arbtt-stats`, and use it together with `--dump-samples`?

**Comment by liskin** [comment]
Just FYI, I ended up doing `arbtt-dump | sed … | jq … | arbtt-import` (https://github.com/liskin/dotfiles/commit/4775aa9e4ff1506ed837d377d536de923b04929b#diff-12f1cea9dd63f36944b5df2ef8cde98eded1a344cb77d16d2c6c6831697131d4R73). It's not particularly fast, but it isn't slow enough to be a problem (takes a couple seconds to process a month worth of data), so I'll probably just stick with it. :-)

**Pull request merged**